File: services/order-service/Infrastructure/Security/RequestContext.cs
namespace order_service.Infrastructure.Security;

public class RequestContext
{
    private readonly IHttpContextAccessor _httpContextAccessor;

    public RequestContext(IHttpContextAccessor httpContextAccessor)
    {
        _httpContextAccessor = httpContextAccessor;
    }

    public string GetUserEmail() => GetRequiredHeader("X-User-Email");
    public string GetUserRole()  => GetRequiredHeader("X-User-Role");
    public string GetUserId()    => GetRequiredHeader("X-User-Id");

    private string GetRequiredHeader(string name)
    {
        var value = _httpContextAccessor.HttpContext?.Request.Headers[name].ToString();
        if (string.IsNullOrWhiteSpace(value))
            throw new InvalidOperationException($"Missing internal header: {name}");
        return value;
    }
}

File: services/inventory-service/src/main/java/com/marko/logistics/inventory/infrastructure/security/RequestContext.java
package com.marko.logistics.inventory.infrastructure.security;

import jakarta.servlet.http.HttpServletRequest;
import org.springframework.stereotype.Component;

@Component
public class RequestContext {

    private static final String HEADER_EMAIL = "X-User-Email";
    private static final String HEADER_ROLE = "X-User-Role";
    private static final String HEADER_ID = "X-User-Id";

    public String getUserEmail(HttpServletRequest request){
        return requireHeader(request, HEADER_EMAIL);
    }

    public String getUserRole(HttpServletRequest request){
        return requireHeader(request, HEADER_ROLE);
    }

    public String getUserId(HttpServletRequest request) {
        return requireHeader(request, HEADER_ID);
    }

    public String requireHeader(HttpServletRequest request, String name){
        String value = request.getHeader(name);
        if(value == null || value.isBlank()) {
            throw new IllegalStateException("Missing internal header: " + name);
        }
        return value;
    }
}